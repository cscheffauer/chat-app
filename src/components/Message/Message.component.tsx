import React from 'react';
import { AppContext } from '../../AppContextProvider';

import Linkify from 'react-linkify';
import './Message.scss';
import { MessageType } from '../../models/chat.model';

interface Props {
	switchToEditMode: (messageid: string, messagetext: string) => void;
	message: MessageType;
}

const Message = ({ switchToEditMode, message }: Props) => {
	const { send, userid } = React.useContext(AppContext);
	const sentDate = new Date(message.sent);
	const minutes = sentDate.getMinutes() < 10 ? '0' + sentDate.getMinutes() : sentDate.getMinutes();
	const hours = sentDate.getHours() < 10 ? '0' + sentDate.getHours() : sentDate.getHours();
	const sent = hours + ':' + minutes;

	const deleteMessage = (messageid: string) => {
		send({ id: messageid, type: 'deletemessageevent' });
	};

	return (
		<div className='message'>
			{message.state === 'DELETED' ? ( //display "Message has been deleted" if message state == DELETED
				<p className='deleted'>Message has been deleted by {message.username}</p>
			) : (
				<>
					<p>
						<span className='sender'>{message.username}</span>
						<span className='info'>{sent}</span>
						{message.state === 'EDITED' && <span className='info'>| edited</span>}
						{message.username !== 'Meetingbot' &&
						message.userid === userid && ( //display edit and delete if message is not from the bot and the username is the same
								<>
									<span className='option' onClick={() => deleteMessage(message.messageid)}>
										×
									</span>
									<span className='option' onClick={() => switchToEditMode(message.messageid, message.text)}>
										✎
									</span>
								</>
							)}
					</p>
					<Linkify>
						<p
							style={{
								lineHeight: '150%',
								color: message.username === 'Meetingbot' ? '#a3a3a3' : 'black',
							}}
						>
							{message.text}
						</p>
					</Linkify>
				</>
			)}
		</div>
	);
};

export default Message;
