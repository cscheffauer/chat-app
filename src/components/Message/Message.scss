@import '../../variables';
.message {
	margin-top: 3vh; /* Fallback for browsers that do not support Custom Properties */
	margin-top: calc(var(--vh, 1vh) * 3);
	margin-left: calc(3% + 5px);
	margin-right: 1%;
	a {
		text-decoration: none;
		color: $color-message-link;
	}
	.sender {
		color: $color-message-sender;
		font-weight: 500;
	}
	.info {
		color: $color-grey;
		margin-left: 10px;
	}
	.option {
		color: $color-grey;
		float: right;
		margin-right: 10px;
		cursor: pointer;
	}
	.deleted {
		color: $color-grey;
		font-weight: 500;
	}
}
