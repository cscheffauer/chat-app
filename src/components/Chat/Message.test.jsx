import React from 'react';
import { mount } from 'enzyme';

import Message from './Message.component';

describe('<Message />', () => {
    const useridMock = "1234";
    const messageidMock = "4321"
    const messageMock = { username: 'Max', text: 'Hallo', sent: new Date('2020-01-01'), userid: useridMock, messageid: messageidMock, state: "" };
    const deletedMock = { username: 'Max', text: 'Hallo', sent: new Date('2020-01-01'), userid: useridMock, messageid: messageidMock, state: "DELETED" };
    const switchToEditModeMock = jest.fn();
    const deleteMessageMock = jest.fn();

    afterEach(() => {
        jest.clearAllMocks();
    });

    const wrapperUserSame = mount(<Message switchToEditMode={switchToEditModeMock} deleteMessage={deleteMessageMock} message={messageMock} userid={useridMock} />)

    it('expect to render Chat component', () => {
        expect(wrapperUserSame).toMatchSnapshot();
    })


    describe('Testing the Message Functions', () => {

        it('should hide message has been deleted information', () => {
            const deleted = wrapperUserSame.find('.deleted');
            expect(deleted.exists()).toBeFalsy();
        });

        it('should delete the message on click', () => {
            const deleteSpan = wrapperUserSame.find('.option').at(0);
            deleteSpan.props().onClick({ preventDefault() { } });

            expect(deleteMessageMock).toHaveBeenCalledWith(messageidMock);
        });

        const wrapperUserDifferent = mount(<Message switchToEditMode={switchToEditModeMock} deleteMessage={deleteMessageMock} message={messageMock} userid={"12"} />)
        it('should hide options if user is different than message userid', () => {
            const options = wrapperUserDifferent.find('.option');
            expect(options.exists()).toBeFalsy();
        });

        const wrapperMessageDeleted = mount(<Message switchToEditMode={switchToEditModeMock} deleteMessage={deleteMessageMock} message={deletedMock} userid={"12"} />)
        it('should show message has been deleted information', () => {
            const deleted = wrapperMessageDeleted.find('.deleted');
            expect(deleted.exists()).toBeTruthy();
            expect(deleted.text()).toEqual("Message has been deleted by " + deletedMock.username)
        });
    });
})


