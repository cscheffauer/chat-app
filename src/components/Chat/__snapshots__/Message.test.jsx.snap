// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> expect to render Chat component 1`] = `
<Message
  deleteMessage={[MockFunction]}
  message={
    Object {
      "messageid": "4321",
      "sent": 2020-01-01T00:00:00.000Z,
      "state": "",
      "text": "Hallo",
      "userid": "1234",
      "username": "Max",
    }
  }
  switchToEditMode={[MockFunction]}
  userid="1234"
>
  <div
    className="message"
  >
    <p>
      <span
        className="sender"
      >
        Max
      </span>
      <span
        className="info"
      >
        01:00
      </span>
      <span
        className="option"
        onClick={[Function]}
      >
        ×
      </span>
      <span
        className="option"
        onClick={[Function]}
      >
        ✎
      </span>
    </p>
    <Linkify
      componentDecorator={[Function]}
      hrefDecorator={[Function]}
      matchDecorator={[Function]}
      textDecorator={[Function]}
    >
      <p
        key="0"
        style={
          Object {
            "color": "black",
            "lineHeight": "150%",
          }
        }
      >
        Hallo
      </p>
    </Linkify>
  </div>
</Message>
`;
